k8s/plugin/src/resources/upgrade.rs

```diff
@@ -0,0 +1,328 @@
+use crate::{
+    constant::{
+        UPGRADE_CONTROLLER_DEPLOYMENT, UPGRADE_IMAGE, UPGRADE_OPERATOR_CLUSTER_ROLE,
+        UPGRADE_OPERATOR_CLUSTER_ROLE_BINDING, UPGRADE_OPERATOR_SERVICE,
+        UPGRADE_OPERATOR_SERVICE_ACCOUNT,
+    },
+    resources::objects,
+};
+
+use anyhow::Error;
+use k8s_openapi::api::{
+    apps::v1::Deployment,
+    core::v1::{Service, ServiceAccount},
+    rbac::v1::{ClusterRole, ClusterRoleBinding},
+};
+use kube::{
+    api::{Api, DeleteParams, ListParams, PostParams},
+    core::ObjectList,
+    Client,
+};
+use std::time::Duration;
+
+/// The types of resources that support the upgrade operator.
+#[derive(clap::Subcommand, Debug)]
+pub enum UpgradeOperator {
+    /// Install, Uninstall upgrade resources.
+    UpgradeOperator,
+}
+
+/// K8s resources needed for upgrade operator
+pub struct UpgradeResources {
+    pub(crate) upgrade_service_account: Api<ServiceAccount>,
+    pub(crate) upgrade_cluster_role: Api<ClusterRole>,
+    pub(crate) upgrade_cluster_role_binding: Api<ClusterRoleBinding>,
+    pub(crate) upgrade_deployment: Api<Deployment>,
+    pub(crate) upgrade_service: Api<Service>,
+}
+
+/// Methods implemented by UpgradesResources
+impl UpgradeResources {
+    /// Returns an instance of UpgradesResources
+    pub async fn new(ns: &str) -> anyhow::Result<Self, Error> {
+        let client = Client::try_default().await?;
+        Ok(Self {
+            upgrade_service_account: Api::<ServiceAccount>::namespaced(client.clone(), ns),
+            upgrade_cluster_role: Api::<ClusterRole>::all(client.clone()),
+            upgrade_cluster_role_binding: Api::<ClusterRoleBinding>::all(client.clone()),
+            upgrade_deployment: Api::<Deployment>::namespaced(client.clone(), ns),
+            upgrade_service: Api::<Service>::namespaced(client, ns),
+        })
+    }
+
+    /// Install the upgrade resources
+    pub async fn install(ns: &str) {
+        match UpgradeResources::new(ns).await {
+            Ok(uo) => {
+                let pp = PostParams::default();
+                // Create a service account
+                let sa = uo.get_service_account().await;
+                if sa.iter().count() == 0 {
+                    let ns: Option<String> = Some(ns.to_string());
+                    let service_account = objects::upgrade_operator_service_account(ns);
+                    let pp = PostParams::default();
+                    match uo
+                        .upgrade_service_account
+                        .create(&pp.clone(), &service_account)
+                        .await
+                    {
+                        Ok(_) => {
+                            println!("service account created");
```

Comment:
I think this log message should specify name and namespace of the serviceaccount.
Same for all of the other resources and error message...
```suggestion
                        Ok(sa) => {
                            println!("ServiceAccount '{}' created in namespace '{}'.", sa.name_any(), sa.metadata.namespace.unwrap());
```